File: coin-change.py
from typing import List

class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        MAX = amount + 1
        table = [MAX] * (amount + 1)
        
        table[0] = 0
        for amt in range(1, amount+1):
            for coin in coins:
                if amt - coin >= 0:
                    table[amt] = min(table[amt], 1 + table[amt - coin])

        return table[amount] if table[amount] != MAX else -1

File: coin-change.cpp
#include <vector>

using namespace std;

class Solution {
public:
  int coinChange(vector<int> &coins, int amount) {
    int MAX = amount + 1;
    vector<int> table(amount + 1, MAX);

    table[0] = 0;
    for (int amt = 1; amt <= amount; ++amt) {
      for (int coin : coins) {
        if (amt - coin >= 0) {
          table[amt] = min(table[amt], 1 + table[amt - coin]);
        }
      }
    }

    return table[amount] == MAX ? -1 : table[amount];
  }
};
